### ISSUE TITLE
C++11 compatibility

### ISSUE THREAD

**Issue by gstonge** [gstonge/SamplableSet#12]
Many people are still using compilers that do not comply with c++17, and are not willing to do the upgrade. Since few features of c++17 are in use, it might be a good idea to make it compatible with c++11.

**Comment by benmaier** [comment]
yes, please! our compute cluster still runs on CentOS 6 and I can't install a C++17-compliant compiler :(

If you point me to the places where C++17 is used as opposed to 11, I could try to help

**Comment by gstonge** [comment]
The only feature we use from C++17 is when we sample. We return a `std::optional`, which allows us to return a `None` when the set is empty. It would be quite simple to change for exceptions instead, I'll look at it tomorrow.

**Comment by benmaier** [comment]
cool! might break people's code though :|

**Comment by gstonge** [comment]
Yeah, that's one of my concerns. I'll start by making a cpp11 branch that uses exceptions, so in the meantime, people can use it with C++11. In the future, I'll try to make it compatible with the current scheme.

**Comment by benmaier** [comment]
hey, as a quick fix, would it be fine to simply remove all instances of "optional" from return types? Given that I would ensure to never sample from an empty set?

**Comment by gstonge** [comment]
Oh sorry about that, I got busy elsewhere and forgot about this issue. Yes, this should work, this is the only major feature I use. I might have used some other c++17 features, like the new map iteration convention, but these can all be made compatible with c++11 easily.

**Comment by gstonge** [comment]
I just pushed the [cpp11](https://github.com/gstonge/SamplableSet/tree/cpp11) branch, which removes `std::optional` and instead raises errors when unallowed actions are made. I'll eventually merge this on the master. Beyond the c++11 compatibility, I feel this is a better convention, and it shouldn't break too much backward compatibility if people made normal use of it (e.g. not sampling empty sets).

If you experience any issue with it let me know!

**Comment by benmaier** [comment]
guillaume I love you
